pkg/controller/infrastructure/infraflow/ensure.go

```diff
@@ -320,12 +320,20 @@ func (fctx *FlowContext) ensureFirewallRules(ctx context.Context) error {
 				log.Info(fmt.Sprintf("failed to create firewall %s rule: %v", rule.Name, err))
 				return err
 			}
-			continue
-		}
-		_, err = fctx.updater.Firewall(ctx, rule)
-		if err != nil {
-			log.Info(fmt.Sprintf("failed to update firewall %s rule: %v", rule.Name, err))
-			return err
+		} else {
+			updateRule, err := firewallUpdate(gcpRule, rule)
+			if err != nil {
+				log.Info(fmt.Sprintf("failed calculate update for firewall %s rule: %v", rule.Name, err))
+				return err
+			}
+			if updateRule != nil {
+				if _, err = fctx.updater.Firewall(ctx, updateRule); err != nil {
```

Comment:
Is it really necessary to compute the diff and update the rule with this diff? Couldn't we just check if we need to update, and if so use `rule` instead of `updateRule`?

Reply:
I'm afraid so.
The entire point of this PR is to _not_ use `rule` as it has _some_ fields set to non-zero values. These fields will cause noop-updates on the GCP side if used in the update call on our side. GCP does not check whether an update is an actual change server-side. The only way around that is to use a firewall-rule with zeroed fields, which is `updateRule`.

Reply:
Not sure I understand the response here. If you detected that the `rule` has not changed, you would skip the update call altogether instead of doing it. If you will do an update anyway, why bother being precise about the fields ?